# SVC model tuned with Grid Search

Training Data Score: 0.28979857050032487

Testing Data Score: 0.24171539961013644

In [1]:
# Dependencies
import pandas as pd
from sklearn.preprocessing import LabelEncoder
import numpy as np
from sklearn import tree
import os

In [2]:
# import processed data

path = "data/"
file = "Best_disposition_data.csv"
path_file = path + file

df = pd.read_csv(path_file)
df = df.drop('Unnamed: 0', axis=1)
df

DATE_year  DATE_month  DATE_week  DATE_day  Hour  Minute  Temp(K)  \
0          2012          10         40         4    20       0   293.31   
1          2012          10         40         4    23       0   291.82   
2          2012          10         40         5     0       1   289.74   
3          2012          10         41        10    14       0   281.87   
4          2012          10         42        15    12      30   292.15   
...         ...         ...        ...       ...   ...     ...      ...   
2047       2017           8         32        11    13       0   293.42   
2048       2017           8         33        20    20      29   304.17   
2049       2017           9         39        27    17      30   297.76   
2050       2017          10         41        13    16      30   289.81   
2051       2017          11         45         9    18      30   282.82   

      Humidity%  Pressure(atm)  wind  Description  disposition  
0          56.0         1017.0   7.0            0            0  
1          68.0         1017.0   3.0           13            2  
2          63.0         1018.0   3.0           13            0  
3          66.0         1020.0   2.0            2            2  
4          82.0         1009.0   8.0            7            0  
...         ...            ...   ...          ...          ...  
2047       88.0         1017.0   2.0           14            4  
2048       54.0         1018.0   1.0           14            2  
2049       73.0         1018.0   5.0            0            2  
2050       67.0         1023.0   2.0           13            2  
2051       70.0         1020.0   5.0           14            2  

[2052 rows x 12 columns]

In [3]:
# Create dataframe for running decision tree
# This was done to upload existing code with similar variables
# Main Data set: data_df

data_df = df.drop("disposition", axis=1)
target = df["disposition"]
#target


feature_names = data_df.columns
data_df.head()

DATE_year  DATE_month  DATE_week  DATE_day  Hour  Minute  Temp(K)  \
0       2012          10         40         4    20       0   293.31   
1       2012          10         40         4    23       0   291.82   
2       2012          10         40         5     0       1   289.74   
3       2012          10         41        10    14       0   281.87   
4       2012          10         42        15    12      30   292.15   

   Humidity%  Pressure(atm)  wind  Description  
0       56.0         1017.0   7.0            0  
1       68.0         1017.0   3.0           13  
2       63.0         1018.0   3.0           13  
3       66.0         1020.0   2.0            2  
4       82.0         1009.0   8.0            7

In [4]:
# Split data into training and testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data_df, target, random_state=42)

In [5]:
# Create the SVC Model
from sklearn.svm import SVC 
model = SVC(kernel='rbf')
model

SVC()

* linear = no

* sigmoid = maybe (<0.4)

* poly = no

* rbf = works well (0.878)

* callable = no

In [6]:
# Create the GridSearch estimator along with a parameter object containing the values to adjust
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [1, 5, 10, 50],
              'gamma': [0.0001, 0.0005, 0.001, 0.005]}
grid = GridSearchCV(model, param_grid, verbose=3)

In [7]:
# Fit the model using the grid search estimator. 
# This will take the SVC model and try each combination of parameters
grid.fit(X_train, y_train)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.231, total=   0.1s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.273, total=   0.1s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.1s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.282, total=   0.1s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.269, total=   0.1s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.270, total=   0.1s
[CV] C=1, gamma=0.0005 ...............................................
[CV] ................... C=1, gamma=0.0005, score=0.250, total=   0.1s
[CV] C=1, gamma=0.0005 ...............................................
[CV] ................... C=1, gamma=0.0005, score=0.292, total=   0.1s
[CV] C=1, gamma=0.0005 ...............................................
[CV] ................... C=1, gamma=0.0005, score=0.279, total=   0.1s
[CV] C=1, gamma=0.0005 ...............................................
[CV] ................... C=1, gamma=0.0005, score=0.292, total=   0.1s
[CV] C=1, gamma=0.0005 ...............................................
[CV] .

[CV] .................. C=50, gamma=0.0001, score=0.286, total=   0.1s
[CV] C=50, gamma=0.0001 ..............................................
[CV] .................. C=50, gamma=0.0001, score=0.266, total=   0.1s
[CV] C=50, gamma=0.0001 ..............................................
[CV] .................. C=50, gamma=0.0001, score=0.292, total=   0.1s
[CV] C=50, gamma=0.0001 ..............................................
[CV] .................. C=50, gamma=0.0001, score=0.244, total=   0.1s
[CV] C=50, gamma=0.0005 ..............................................
[CV] .................. C=50, gamma=0.0005, score=0.240, total=   0.2s
[CV] C=50, gamma=0.0005 ..............................................
[CV] .................. C=50, gamma=0.0005, score=0.218, total=   0.2s
[CV] C=50, gamma=0.0005 ..............................................
[CV] .................. C=50, gamma=0.0005, score=0.244, total=   0.2s
[CV] C=50, gamma=0.0005 ..............................................
[CV] .

[Parallel(n_jobs=1)]: Done  80 out of  80 | elapsed:    9.9s finished


GridSearchCV(estimator=SVC(),
             param_grid={'C': [1, 5, 10, 50],
                         'gamma': [0.0001, 0.0005, 0.001, 0.005]},
             verbose=3)

In [8]:
# List the best parameters for this dataset
print(grid.best_params_)

{'C': 1, 'gamma': 0.0005}


In [9]:
# List the best score
print(grid.best_score_)

0.2774503997631033


In [10]:
print('Test Acc: %.3f' % grid.score(X_test, y_test))

Test Acc: 0.250


In [11]:
print(f"Training Data Score: {grid.score(X_train, y_train)}")
print(f"Testing Data Score: {grid.score(X_test, y_test)}")

Training Data Score: 0.2969460688758934
Testing Data Score: 0.24951267056530213


In [12]:
# Make predictions with the hypertuned model
predictions = grid.predict(X_test)

In [13]:
optimized_model = grid.predict(X_test)

In [14]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions,
                            target_names=["DT", "ND", "NE","NW","SE","SW"]))

              precision    recall  f1-score   support

          DT       0.26      0.83      0.39       121
          ND       0.00      0.00      0.00        81
          NE       0.22      0.27      0.24       104
          NW       0.00      0.00      0.00        62
          SE       0.00      0.00      0.00        81
          SW       0.00      0.00      0.00        64

    accuracy                           0.25       513
   macro avg       0.08      0.18      0.11       513
weighted avg       0.11      0.25      0.14       513



C:\Users\phoog\anaconda3\envs\PythonData\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [15]:
#optimized_model= grid.predict(X_test)

#Save the model
#optimized_model.save("svm_model_grid_search.h5")

In [16]:
#optimized model 

grid.predict(X_test)

array([2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 2, 0, 0, 2, 2, 0, 0, 2, 0,
       0, 0, 0, 0, 0, 2, 0, 0, 2, 2, 5, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0,
       2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 2, 2, 0,
       0, 0, 2, 0, 0, 0, 2, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 2, 2, 0, 2, 2, 0, 0, 0, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 2, 2, 0, 0, 2, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 2, 0, 0, 2, 0, 0,
       0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2, 0, 0, 0,
       0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0, 0, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 2, 0, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 0, 2, 2, 2, 0,
       0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 2, 0, 0,

Making Predictions with new data
Label asignments

**DT:0

**NE:2

**NW:3

**SE:4

**SW:5

DATE_year DATE_month DATE_week DATE_day Hour Minute Temp(K) Humidity% Pressure(atm) wind Street (Coded) 2012 10 40 4 20 0 293.31 56 1017 7 1675

2012 12 49 7 23 0 281.34 70 1016 2 1620

2017 7 28 6 13 27 292.06 88 1017 1 618

SET 1 ANSWER 0

SET 2 ANSWER 2

SET 3 ANSER 0

In [17]:
new_data =   np.array([[2012,12, 49, 7, 23, 0, 281.34, 70, 1016, 2]])
new_data_2 = np.array([[2017,7, 28, 6, 13, 27, 292.06, 88, 1017, 1]]) 

In [18]:
# Test the model
#new_data = np.array([[0, 0, 0, 25.8, 2.83, 0, 0.11600, -0.19, 0.32, -0.11600]])
print(f"Predicted class: {grid.predict(new_data)}")

ValueError: X.shape[1] = 10 should be equal to 11, the number of features at training time

In [ ]:
 # save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'z2_SVC_tuned_gridSearch_model.sav'
joblib.dump(optimized_model, filename)